addons/giltracker/giltracker.lua

```diff
@@ -0,0 +1,212 @@
+        SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+]]
+
+_addon.name = 'giltracker'
+_addon.author = 'sylandro'
+_addon.version = '1.0.0'
+_addon.language = 'English'
+
+config = require('config')
+images = require('images')
+texts = require('texts')
+
+local GIL_ITEM_ID = 0xFFFF
+local CUTSCENE_STATUS_ID = 4
+local SCROLL_LOCK_KEY = 70
+
+hideKey = SCROLL_LOCK_KEY
+is_hidden_by_cutscene = false
+is_hidden_by_key = false
+
+defaults = {}
+defaults.hideKey = SCROLL_LOCK_KEY
+defaults.gilText = {}
+defaults.gilText.bg = {}
+defaults.gilText.bg.alpha = 100
+defaults.gilText.bg.red = 0
+defaults.gilText.bg.green = 0
+defaults.gilText.bg.blue = 0
+defaults.gilText.bg.visible = false
+defaults.gilText.text = {}
+defaults.gilText.text.font = 'sans-serif'
+defaults.gilText.text.fonts = {'Arial','Trebuchet MS'}
+defaults.gilText.text.size = 9
+defaults.gilText.flags = {}
+defaults.gilText.flags.italic = true
+defaults.gilText.flags.bold = false
+defaults.gilText.flags.right = true
+defaults.gilText.flags.bottom = true
+defaults.gilText.pos = {}
+defaults.gilText.pos.x = -285
+defaults.gilText.pos.y = -35
+defaults.gilText.text.alpha = 255
+defaults.gilText.text.red = 253
+defaults.gilText.text.green = 252
+defaults.gilText.text.blue = 250
+defaults.gilText.text.stroke = {}
+defaults.gilText.text.stroke.alpha = 200
+defaults.gilText.text.stroke.red = 50
+defaults.gilText.text.stroke.green = 50
+defaults.gilText.text.stroke.blue = 50
+defaults.gilText.text.stroke.width = 2
+defaults.gilText.text.visible = true
+defaults.gilImage = {}
+defaults.gilImage.color = {}
+defaults.gilImage.color.alpha = 255
+defaults.gilImage.color.red = 255
+defaults.gilImage.color.green = 255
+defaults.gilImage.color.blue = 255
+defaults.gilImage.visible = true
+
+local settings = config.load(defaults)
+config.save(settings)
+
+settings.gilImage.texture = {}
+settings.gilImage.texture.path = windower.addon_path..'gil.png'
+settings.gilImage.texture.fit = true
+settings.gilImage.size = {}
+settings.gilImage.size.height = 23
+settings.gilImage.size.width = 23
+settings.gilImage.draggable = false
+settings.gilImage.repeatable = {}
+settings.gilImage.repeatable.x = 1
+settings.gilImage.repeatable.y = 1
+
+gil_image = images.new(settings.gilImage)
+gil_text = texts.new(settings.gilText)
+
+config.register(settings, function(settings)
+    hideKey = settings.hideKey
+end)
+
+windower.register_event('load',function()
+    if windower.ffxi.get_info().logged_in then
+        initialize()
+    end
+end)
+
+windower.register_event('login', function()
+    initialize()
+end)
+
+windower.register_event('logout', function(...)
+    hide()
+end)
+
+windower.register_event('add item', function(...)
+    update_gil_if_item_id_matches(id)
+end)
+
+windower.register_event('remove item', function(bag, slot, id, count)
+    update_gil_if_item_id_matches(id)
+end)
+
+windower.register_event('incoming text', function(original, ...)
```

Comment:
Why are you checking for this event? Are there situations where the first two events (add/remove item) don't cover all the cases? Even if so, I would update the matching function, yours updates every time the characters `g`, `i` and `l` appear next to each other. So any time someone mentions "agility", "gilded", "fragile", etc. it will update the gil count. Which isn't *bad*, just unnecessary.

Reply:
For some reason, when you receive gil when you defeat a mob the 'add item' event doesn't get triggered, hence the need to read the 'gil' part of an incoming text to trigger the update of the gil counter. I can't think of a better solution to this...

It is strange that the add item event doesn't fire, though.

Reply:
Gil should *always* be updated by one of the incoming item packets, which should appropriately call the `add item` or `remove item` events. If you do ever notice a discrepancy, you can make an issue though, because that would mean there's a problem!

Reply:
I'll make an issue, I definitely have been able to reproduce the problem.

There is another case when gil doesn't get updated by incoming packets: during login. I traced it with logs and item 0xFFFF is never added. For now, I'll handle it with the "Stop downloading" packet. I suppose that there's another packet that updates currencies or something?

Reply:
If you use PacketViewer you can check your current gil amount, log out, then type this into the console:
```
packetviewer scan incoming 123456
```

Then log in, it will scan all incoming packets for this value. That should tell you which packet carries that information. It's possible (and wouldn't surprise me at all) if the gil value is also transmitted separately from other items.

However, it's also possible that it just isn't transmitted at all - after all you don't need to know it until you open the menu, which sends a packet that requests some info from the server (like player stats). If that turns out to be the case, you also force that info by injecting that outgoing packet on login.